tools/src/wyvern/tools/typedAST/core/declarations/DefDeclaration.java

```diff
@@ -235,29 +254,82 @@ public Environment extendName(Environment env, Environment against) {
 	@Override
 	public DeclType genILType(GenContext ctx) {
 		List<FormalArg> args = new LinkedList<FormalArg>();
-		for (NameBinding b : argNames) {
+
+        ctx = this.serializeArguments(args, ctx);
+
+		DefDeclType ret = new DefDeclType(getName(), getResultILType(ctx), args);
+		return ret;
+	}
+
+    private GenContext serializeArguments(List<FormalArg> args, GenContext ctx) {
+        if(isGeneric()) {
+            ValueType type = this.genericStructuralType();
+            args.add(new FormalArg(GENERIC_PREFACE, type));
+            ctx = ctx.extend(GENERIC_PREFACE, new Variable(GENERIC_PREFACE), type);
```

Comment:
instead, use:

`ctx = new TypeGenContext(genTypeName, GENERIC_PREFACE, ctx);`

This will automatically translate all references to genTypeName as a type into GENERIC_PREFACE.genTypeName